ide/coqide/coqOps.ml

```diff
@@ -494,11 +494,14 @@ object(self)
 
   method private attach_tooltip ?loc sentence text =
     let start_sentence, stop_sentence, phrase = self#get_sentence sentence in
-    let pre_bytes, post_bytes = Option.cata Loc.unloc (0, String.length phrase) loc in
-    let pre = b2c phrase (pre_bytes - (sentence.byte_start)) in
-    let post = b2c phrase (post_bytes - (sentence.byte_start)) in
-    let start = start_sentence#forward_chars pre in
-    let stop = start_sentence#forward_chars post in
+    let start, stop = match loc with
+      | None -> start_sentence, stop_sentence
+      | Some loc ->
+        let start, stop = new GText.iter (start_sentence#as_iter), new GText.iter (start_sentence#as_iter) in
+        let start = start#set_offset loc.Loc.bp in
+        let stop = stop#set_offset loc.ep in
```

Comment:
Losing the `b2c` here is not correct when there are unicode characters in the sentence.  The test case in #16987 only triggers the `None` case.  The issue is that `Loc` uses byte offsets while `buffer` uses unicode character offsets, so a conversion is necessary, else you may highlight too much.  (Also add more stuff after the sentence getting the warning for the test.)

This seems to be what's needed in master, though I have to test it some more.
```
        let pre = b2c phrase loc.Loc.bp in
        let post = b2c phrase loc.ep in
```

I'll get back to this over the weekend.

Reply:
I will submit a separate PR to fix the warning tooltips in a separate PR for 8.17 shortly.  The case where `loc` is `None` is the important one--none of the calls to `CWarnings.create` pass loc.  I haven't figured out the fix for the `Some` case yet.  IMO that's not urgent, though I will work on it in the next few days.

Error messages also generate tooltips.  These seem to work correctly even when `loc` is provided.  FWIW, I found 79 places in the source that can provide `loc`.

Reply:
Solved thanks; used GTK native methods instead of `b2c` now that Coq returns the full location info correctly.